File: ios/BiblosSurfer/Tools/UITestStubLibraryService.swift
//
//  UITestStubLibraryService.swift
//  BiblosSurfer
//
//  Created by Mikołaj Linczewski on 21/08/2026.
//

import Foundation

/// Deterministic library for previews and UI tests. Points at the bundled Genesis EPUB so the
/// screens never depend on files in Documents.
final class UITestStubLibraryService: LibraryServiceProtocol {
    static let sampleTitle = "Genesis. Księga Rodzaju. Bereszit"
    static let sampleAuthor = "Izaak Cylkow"
    static let resourceName = "genesis-ksiega-rodzaju-bereszit"
    static let fileExtension = "epub"

    var sampleBookURL: URL? {
        Bundle.main.url(forResource: Self.resourceName, withExtension: Self.fileExtension)
    }

    func loadItems() async throws -> [LibraryItem] {
        guard let sampleBookURL else {
            throw Errors.Library.fileMissing(fileName: "\(Self.resourceName).\(Self.fileExtension)")
        }
        return [
            LibraryItem(
                id: UUID(uuidString: "AAAAAAAA-BBBB-CCCC-DDDD-EEEEEEEEEEEE") ?? UUID(),
                fileName: "\(Self.resourceName).\(Self.fileExtension)",
                title: Self.sampleTitle,
                author: Self.sampleAuthor,
                fileURL: sampleBookURL,
                coverURL: nil,
                locatorJSON: nil,
                progression: 0,
                format: .epub,
                folderName: nil
            )
        ]
    }

    func importBook(from sourceURL: URL) async throws -> LibraryItem {
        throw Errors.Library.unsupportedFormat(fileExtension: sourceURL.pathExtension)
    }
}

File: android/app/src/main/java/miko/biblossurfer/data/UITestStubLibraryService.kt
package miko.biblossurfer.data

import android.content.Context
import miko.biblossurfer.data.model.LibraryItem
import miko.biblossurfer.data.model.PublicationFormat
import java.util.UUID

/**
 * Deterministic library for previews and UI tests. Points at the bundled Genesis EPUB so the
 * screens never depend on files in app storage.
 */
class UITestStubLibraryService(private val context: Context) : LibraryServiceProtocol {
    companion object {
        const val sampleTitle = "Genesis. Księga Rodzaju. Bereszit"
        const val sampleAuthor = "Izaak Cylkow"
        const val resourceName = "genesis-ksiega-rodzaju-bereszit"
        const val fileExtension = "epub"
    }

    val sampleBookFile = context.bundledEpubFiles()
        .firstOrNull { it.name == "$resourceName.$fileExtension" }

    override suspend fun loadItems(): List<LibraryItem> {
        val file = sampleBookFile
            ?: throw Errors.Library.FileMissing("$resourceName.$fileExtension")
        return listOf(
            LibraryItem(
                id = UUID.fromString("AAAAAAAA-BBBB-CCCC-DDDD-EEEEEEEEEEEE"),
                fileName = "$resourceName.$fileExtension",
                title = sampleTitle,
                author = sampleAuthor,
                fileURL = file,
                coverURL = null,
                locatorJSON = null,
                progression = 0.0,
                format = PublicationFormat.EPUB,
                folderName = null,
            )
        )
    }

    override suspend fun importBook(from: java.io.File): LibraryItem {
        throw Errors.Library.UnsupportedFormat(from.extension)
    }
}
